app/overnight/purge-soft-deleted-records.js

```diff
@@ -0,0 +1,152 @@
+ * @name DeletedRecordCount#total
+ * @type {number}
+ * @readonly
+ */
+Object.defineProperty(DeletedRecordCount.prototype, 'total', {
+  get () {
+    return this.dogs + this.owners
+  }
+})
+
+/**
+ * @constructor
+ */
+function DeletedRecords () {
+  /**
+   * @type {string[]}
+   */
+  this.dogs = []
+  /**
+   * @type {string[]}
+   */
+  this.owners = []
+}
+
+/**
+ * @typedef {{
+ *  count: {
+ *    success: DeletedRecordCount,
+ *    failed: DeletedRecordCount
+ *  },
+ *  deleted: {
+ *    success: DeletedRecords,
+ *    failed: DeletedRecords
+ *  }
+ * }} SoftDeletedRecordPayload
+ */
+
+/**
+ * @param {Date} date
+ * @return {Promise<SoftDeletedRecordPayload>}
+ */
+const purgeSoftDeletedRecords = async (date = new Date()) => {
+  try {
+    const deleteAtCutOff = new Date(date)
+    deleteAtCutOff.setUTCDate(deleteAtCutOff.getUTCDate() - paranoidRetentionPeriod)
+
+    const ownerResults = await sequelize.models.person.findAll({
+      where: {
+        deleted_at: {
+          [Op.lte]: deleteAtCutOff
+        }
+      },
+      paranoid: false
+    })
+
+    /**
+     * @type {SoftDeletedRecordPayload}
+     */
+    const result = {
+      count: {
+        success: new DeletedRecordCount(),
+        failed: new DeletedRecordCount()
+      },
+      deleted: {
+        success: new DeletedRecords(),
+        failed: new DeletedRecords()
+      },
+      toString () {
+        return 'Purge deleted records.' +
+          ` Success: ${this.count.success.dogs} Dogs ${this.count.success.owners} Owners - [${[...this.deleted.success.dogs, ...this.deleted.success.owners].join(', ')}].` +
+          ` Failed: ${this.count.failed.dogs} Dogs ${this.count.failed.owners} Owners - [${[...this.deleted.failed.dogs, ...this.deleted.failed.owners].join(', ')}].`
+      }
+    }
+
+    const dogResults = await sequelize.models.dog.findAll({
+      where: {
+        deleted_at: {
+          [Op.lte]: deleteAtCutOff
+        }
+      },
+      paranoid: false
+    })
+
+    for (const dog of dogResults) {
+      try {
+        await purgeDogByIndexNumber(dog.index_number, overnightJobUser)
+        result.deleted.success.dogs.push(dog.index_number)
+        result.count.success.dogs++
+      } catch (e) {
+        console.log(`Failed to hard delete dog record ${dog.index_number}`, e)
+        result.count.failed.dogs++
+      }
+    }
+
+    for (const owner of ownerResults) {
+      try {
+        await purgePersonByReferenceNumber(owner.person_reference, overnightJobUser)
+        result.deleted.success.owners.push(owner.person_reference)
+        result.count.success.owners++
+      } catch (e) {
+        console.log(`Failed to hard delete owner record ${owner.person_reference}`, e)
+        result.count.failed.owners++
+      }
+    }
+
+    return result
+  } catch (e) {
+    console.log('Error purging soft deleted records:', e)
+    return {
+      count: {
+        success: new DeletedRecordCount(),
```

Comment:
Is this not clearing the stats of how many success/failures so far?

Reply:
this branch is only traversed when there is an uncaught exception, which could most likely occur on line 71. As such will not have any counts.  Other possible exceptions are caught in the try catch blocks which increment the stats.